src/arkode/arkode_io.c

```diff
@@ -819,389 +877,220 @@ int arkSetMaxNumConstrFails(void *arkode_mem, int maxfails)
-int arkSetFixedStepBounds(void *arkode_mem, realtype lb, realtype ub)
-{
-  int retval;
-  ARKodeHAdaptMem hadapt_mem;
-  ARKodeMem ark_mem;
-  retval = arkAccessHAdaptMem(arkode_mem, "arkSetFixedStepBounds",
-                              &ark_mem, &hadapt_mem);
-  if (retval != ARK_SUCCESS)  return(retval);
-
-  /* set allowable interval, otherwise set defaults */
-  if ((lb <= ONE) && (ub >= ONE)) {
-    hadapt_mem->lbound = lb;
-    hadapt_mem->ubound = ub;
-  } else {
-    hadapt_mem->lbound = HFIXED_LB;
-    hadapt_mem->ubound = HFIXED_UB;
-  }
-
-  return(ARK_SUCCESS);
-}
-
-
-/*---------------------------------------------------------------
-  arkSetAdaptivityMethod:
+  arkSetAdaptivityMethod:  ***DEPRECATED***
 
   Specifies the built-in time step adaptivity algorithm (and
   optionally, its associated parameters) to use.  All parameters
   will be checked for validity when used by the solver.
+
+  Users should transition to constructing non-default SUNControl
+  objects directly, and providing those directly to the integrator
+  via the time-stepping module *SetController routines.
   ---------------------------------------------------------------*/
 int arkSetAdaptivityMethod(void *arkode_mem, int imethod, int idefault,
                            int pq, realtype adapt_params[3])
 {
   int retval;
-  ARKodeHAdaptMem hadapt_mem;
+  long int lenrw, leniw;
+  realtype k1, k2, k3;
   ARKodeMem ark_mem;
-  retval = arkAccessHAdaptMem(arkode_mem, "arkSetAdaptivityMethod",
-                              &ark_mem, &hadapt_mem);
-  if (retval != ARK_SUCCESS)  return(retval);
+  if (arkode_mem==NULL) {
+    arkProcessError(NULL, ARK_MEM_NULL, "ARKODE",
+                    "arkSetController", MSG_ARK_NO_MEM);
+    return(ARK_MEM_NULL);
+  }
+  ark_mem = (ARKodeMem) arkode_mem;
+
+  /* Remove current SUNControl object */
+  retval = SUNControlSpace(ark_mem->hcontroller, &lenrw, &leniw);
+  if (retval == SUNCONTROL_SUCCESS) {
+    ark_mem->liw -= leniw;
+    ark_mem->lrw -= lenrw;
+  }
+  SUNControlDestroy(ark_mem->hcontroller);
+  ark_mem->hcontroller = NULL;
 
-  /* check for allowable parameters */
-  if ((imethod > ARK_ADAPT_IMEX_GUS) || (imethod < ARK_ADAPT_PID)) {
+  /* set adaptivity parameters from inputs or signal use of defaults */
+  if (idefault == 1) {
+    k1 = -RCONST(1.0);
+    k2 = -RCONST(1.0);
+    k3 = -RCONST(1.0);
+  } else {
+    k1 = adapt_params[0];
+    k2 = adapt_params[1];
+    k3 = adapt_params[2];
+  }
+
+  /* Create new SUNControl object based on "imethod" input, optionally setting
+     the specified controller parameters */
+  SUNControl C = NULL;
+  switch (imethod) {
+  case (ARK_ADAPT_PID):
+    C = SUNControlPID(ark_mem->sunctx);
+    if (C == NULL) {
+      arkProcessError(ark_mem, ARK_MEM_FAIL, "ARKODE", "arkSetAdaptivityMethod",
+                      "SUNControlPID allocation failure");
+      return(ARK_MEM_FAIL);
+    }
+    retval = SUNControlPID_SetParams(C, pq, k1, k2, k3);
```

Comment:
Here and for the other methods below, should this only be called if `adapt_params != NULL`?